File: lib/plotting_r/stages/peak_calling.R
# R-side peak_calling stage: profile_pileup, coverage_track,
# peak_saturation. Mirrors lib/plotting/stages/peak_calling.py.
#
# Plan reference: §S13.6 — R parity for chip-seq + atac-seq peak
# calling figures. Each figure resolves a manifest table path and
# delegates to the matching `ecaa_*_r` primitive in core.R.

if (!exists("ecaa_register_figure")) {
  stop("source runtime/plotting_r/core.R before this stage module")
}

ecaa_register_figure("peak_calling", "profile_pileup", function(ctx) {
  p <- .ecaa_manifest_path(ctx$manifest, ctx$outputs_dir, "profile_table")
  if (is.null(p)) stop("manifest.profile_table required")
  df <- .ecaa_load_tsv(p, list(
    position = c("position", "distance", "offset"),
    signal = c("signal", "coverage", "rpkm", "fold_enrichment"),
    `group?` = c("group", "antibody", "sample", "condition")
  ))
  if (is.null(df)) stop(sprintf("unparseable profile table: %s", p))
  ecaa_profile_pileup_r(df, title = "Profile pileup",
                        group_col = if ("group" %in% colnames(df)) "group" else NULL)
})

ecaa_register_figure("peak_calling", "coverage_track", function(ctx) {
  p <- .ecaa_manifest_path(ctx$manifest, ctx$outputs_dir, "coverage_table")
  if (is.null(p)) stop("manifest.coverage_table required")
  df <- .ecaa_load_tsv(p, list(
    chrom = c("chrom", "chromosome", "#chrom"),
    pos = c("pos", "position"),
    depth = c("depth", "coverage", "n_reads")
  ))
  if (is.null(df)) stop(sprintf("unparseable coverage table: %s", p))
  ecaa_coverage_track_r(df, title = "Coverage track")
})

ecaa_register_figure("peak_calling", "peak_saturation", function(ctx) {
  p <- .ecaa_manifest_path(ctx$manifest, ctx$outputs_dir, "saturation_table")
  if (is.null(p)) stop("manifest.saturation_table required")
  df <- .ecaa_load_tsv(p, list(
    depth = c("depth", "subsampled_reads", "n_reads"),
    peaks_called = c("peaks_called", "n_peaks", "peaks"),
    `group?` = c("group", "sample", "replicate")
  ))
  if (is.null(df)) stop(sprintf("unparseable saturation table: %s", p))
  ecaa_peak_saturation_r(df, title = "Peak saturation",
                         group_col = if ("group" %in% colnames(df)) "group" else NULL)
})

File: testdata/emitted-packages/14-bouwmeester-lfq-proteomics/runtime/plotting/stages/peak_calling.py
"""Peak calling stage — chip-seq + atac-seq taxonomies. Renders the
profile pileup around peak summits, an IGV-style coverage track for a
representative locus, and a saturation curve to confirm the read
depth was sufficient.

Plan reference: §S13 (chip-seq + atac-seq taxonomy stage `peak_calling`).

Manifest contract:
- ``profile_table``: TSV with `position`, `signal`, optional `group`.
- ``coverage_table``: TSV with `chrom`, `pos`, `depth`.
- ``saturation_table``: TSV with `depth`, `peaks_called`, optional `group`.
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional

from ..core import (
    FigureContext,
    coverage_track,
    peak_saturation,
    profile_pileup,
    register_figure,
    stage_registry,
)
from ._shared import load_tsv_columns, manifest_path

FIGURES = stage_registry("peak_calling")


@register_figure(FIGURES, "profile_pileup")
def profile_pileup_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "profile_table")
    if p is None:
        raise FileNotFoundError("manifest.profile_table required")
    cols = load_tsv_columns(
        p,
        {
            "position": ("position", "distance", "offset"),
            "signal": ("signal", "coverage", "rpkm", "fold_enrichment"),
            "group?": ("group", "antibody", "sample", "condition"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable profile table: {p}")
    group_col = "group" if "group" in cols else None
    return profile_pileup(
        frame=cols,
        title="Profile pileup",
        out=out,
        group_col=group_col,
    )


@register_figure(FIGURES, "coverage_track")
def coverage_track_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "coverage_table")
    if p is None:
        raise FileNotFoundError("manifest.coverage_table required")
    cols = load_tsv_columns(
        p,
        {
            "chrom": ("chrom", "chromosome", "#CHROM"),
            "pos": ("pos", "position", "POS"),
            "depth": ("depth", "coverage", "n_reads"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable coverage table: {p}")
    return coverage_track(frame=cols, title="Coverage track", out=out)


@register_figure(FIGURES, "peak_saturation")
def peak_saturation_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "saturation_table")
    if p is None:
        raise FileNotFoundError("manifest.saturation_table required")
    cols = load_tsv_columns(
        p,
        {
            "depth": ("depth", "subsampled_reads", "n_reads"),
            "peaks_called": ("peaks_called", "n_peaks", "peaks"),
            "group?": ("group", "sample", "replicate"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable saturation table: {p}")
    group_col = "group" if "group" in cols else None
    return peak_saturation(
        frame=cols,
        title="Peak saturation",
        out=out,
        group_col=group_col,
    )
